File: 39-combination-sum/39-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res= new LinkedList<>();
        combinationSum(candidates, 0, target, new LinkedList<>(), res);
        return res;
    }
    public void combinationSum(int[] candidates, int i, int target, List<Integer>cur, List<List<Integer>> res) {
        
        if(target <= 0){
            
            if(target == 0)
                res.add(new LinkedList<>(cur));
            return;
        }
        for(; i < candidates.length; i++){
            
            cur.add(candidates[i]);
            combinationSum(candidates, i, target - candidates[i], cur, res);
            cur.remove(cur.size() - 1);            
        }
    
    }    
}

File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>res;
        vector<int>cur;
        combinationSum(res, cur, candidates, target, 0);
        return res;
    }
    void combinationSum(vector<vector<int>>& res, vector<int> cur, vector<int>& candidates, int target, int i){
        if(target <= 0){
            if(target == 0){
                res.push_back(cur);                
            }
            return;
        }
        for(; i < candidates.size(); i++){
            cur.push_back(candidates[i]);
            combinationSum(res, cur, candidates, target - candidates[i], i);
            cur.pop_back();
        }
    }
};